[memprof] Add Version2 of IndexedMemProfRecord serialization

kazutakahirata opened the pull request:
I'm currently developing a new version of the indexed memprof format
where we deduplicate call stacks in IndexedAllocationInfo::CallStack
and IndexedMemProfRecord::CallSites.  We refer to call stacks with
integer IDs, namely CallStackId, just as we refer to Frame with
FrameId.  The deduplication will cut down the profile file size by 80%
in a large memprof file of mine.

As a step toward the goal, this patch teaches
IndexedMemProfRecord::{serialize,deserialize} to speak Version2.  A
subsequent patch will add Version2 support to llvm-profdata.

The essense of the patch is to replace the serialization of a call
stack, a vector of FrameIDs, with that of a CallStackId.  That is:

  const IndexedAllocationInfo &N = ...;
  ...
  LE.write<uint64_t>(N.CallStack.size());
  for (const FrameId &Id : N.CallStack)
    LE.write<FrameId>(Id);

becomes:

  LE.write<CallStackId>(N.CSId);

kazutakahirata committed:
[memprof] Add Version2 of IndexedMemProfRecord serialization

I'm currently developing a new version of the indexed memprof format
where we deduplicate call stacks in IndexedAllocationInfo::CallStack
and IndexedMemProfRecord::CallSites.  We refer to call stacks with
integer IDs, namely CallStackId, just as we refer to Frame with
FrameId.  The deduplication will cut down the profile file size by 80%
in a large memprof file of mine.

As a step toward the goal, this patch teaches
IndexedMemProfRecord::{serialize,deserialize} to speak Version2.  A
subsequent patch will add Version2 support to llvm-profdata.

The essense of the patch is to replace the serialization of a call
stack, a vector of FrameIDs, with that of a CallStackId.  That is:

  const IndexedAllocationInfo &N = ...;
  ...
  LE.write<uint64_t>(N.CallStack.size());
  for (const FrameId &Id : N.CallStack)
    LE.write<FrameId>(Id);

becomes:

  LE.write<CallStackId>(N.CSId);

teresajohnson commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -357,6 +372,9 @@ struct IndexedMemProfRecord {
   // the last entry in the list with the same function GUID.
   llvm::SmallVector<llvm::SmallVector<FrameId>> CallSites;
+  // Conceptually the same as above.  We are going to keep both CallSites and
+  // CallSiteIds while we are transitioning from CallSites to CallSitesIds.
```
s/CallSitesIds/CallSiteIds/

teresajohnson commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -306,6 +318,9 @@ struct IndexedAllocationInfo {
         return false;
     }
+
+    if (Other.CSId != CSId)
```
If CSId should (already) be in sync with the CallStack, can the above CallStack comparisons be removed now?

kazutakahirata replied:
Yes, we can remove the `CallStack` comparison.  I was planning to remove it when we remove the `CallStack` field, but we can do so now.

teresajohnson commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -401,16 +424,21 @@ struct IndexedMemProfRecord {
         return false;
     }
+
+    if (Other.CallSiteIds != CallSiteIds)
```
Ditto to same question in IndexedAllocationInfo operator==: can the CallSites based comparison above be removed?

kazutakahirata replied:
Yes, we can remove the `CallSites` comparison.  I was planning to remove it when we remove the `CallSites` field, but we can do so now.

teresajohnson commented on llvm/unittests/ProfileData/MemProfTest.cpp:
```diff
@@ -265,7 +265,7 @@ TEST(MemProf, PortableWrapper) {
 }
 
-TEST(MemProf, RecordSerializationRoundTrip) {
+TEST(MemProf, RecordSerializationRoundTripVersion0) {
```
Should we also have a test for version 1?

kazutakahirata replied:
Maybe I'll rename the test to `Version0AndVerion1` or something because the serialization format is the same between those two versions.

teresajohnson replied:
Oh I overlooked the fact that the serialization format of these records is not changed between versions 0 and 1. So another option is to simply add a comment here that version 1 is the same.

kazutakahirata replied:
Oops, I read your comment after I updated the patch, but, yes, I did put a comment there.

teresajohnson commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -289,10 +291,20 @@ struct IndexedAllocationInfo {
-    return sizeof(uint64_t) + // The number of frames to serialize.
-           sizeof(FrameId) * CallStack.size() +    // The callstack frame ids.
-           PortableMemInfoBlock::serializedSize(); // The size of the payload.
+  size_t serializedSize(IndexedVersion Version = Version0) const {
```
For here and all other functions below that currently default this parameter to Version0: Since the current version is Version1, should that be the default? Or better - should Version be a required parameter?

kazutakahirata replied:
I was on the fence whether to use a required parameter or `= Version0`, but I chose to use `= Version0` to keep the patch small.  I'll switch to the required parameter.

teresajohnson replied:
Or would it make sense to have a variable that tracks the current version and default to that?

kazutakahirata replied:
I have a slight preference to telling these functions (`serialize`, `serializedSize`, `deserialize`) exactly which version to use and leaving any smarts higher up.  This way, a very small number of places know what the current version is.

teresajohnson reviewed (commented):
> A subsequent patch will add Version2 support to llvm-profdata.

Do you generally mean the reader / writer?

kazutakahirata commented:
> > A subsequent patch will add Version2 support to llvm-profdata.
> 
> Do you generally mean the reader / writer?

Yes.

kazutakahirata committed:
Removed CallStack-based comparison.

Removed CallSites-based comparison.

Made Version a required parameter in
serialize/deserialize/serializedSize.

kazutakahirata committed:
Fix a typo.

kazutakahirata commented:
I believe I've addressed all the comments.  Please take a look at the latest iteration.  Thanks!

teresajohnson commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -535,7 +519,7 @@ class RecordLookupTrait {
   data_type ReadData(uint64_t K, const unsigned char *D,
                      offset_type /*Unused*/) {
-    Record = IndexedMemProfRecord::deserialize(Schema, D);
+    Record = IndexedMemProfRecord::deserialize(Schema, D, Version1);
```
Suggest having a CurrentVersion variable in the header where the versions are defined and set it to Version1 and use here and elsewhere. Will greatly simplify updating to the new version.

kazutakahirata replied:
When I teach the reader/writer to use `Version2`, I'm planning to take a version as a constructor parameter:

```
 RecordLookupTrait(IndexedVersion Version, const MemProfSchema &S)
     : Version(Version), Schema(S) {}
```

and pass `Version` to `IndexedMemProfRecord::deserialize` here.  Is that OK?

teresajohnson replied:
Maybe just put part of that in this patch, since it would be confined to this same file - i.e. have a member of RecordLookupTrait that you initialize to Version1 and pass down via deserialize. Ditto I guess for RecordWriterTrait and serialize()/serializedSize().

kazutakahirata replied:
I've included the changes to `RecordLookupTrait` and `RecordWriterTrait` in the latest iteration.

snehasish reviewed (commented):
I wonder if it makes sense to have a higher level distinction of versioning instead of pushing the condition down into each method. I'm trying to think about how difficult it will be in the future to evolve to the next version. For example, say we have `serializedSizeV1()` as the old implementation and `serializedSizeV2()` as the new implementation and `serializedSize()` calls one or the other based on the version. Pushing it down one level makes it easy to just delete the old function when we introduce v3. In this case the difference is not large since the logic varies only a little but it may make sense for more complex code. What do you both think?

kazutakahirata committed:
Use separate functions for different versions.

kazutakahirata commented:
> I wonder if it makes sense to have a higher level distinction of versioning instead of pushing the condition down into each method. I'm trying to think about how difficult it will be in the future to evolve to the next version. For example,
> 
> Say we have `serializedSizeV1()` as the old implementation and `serializedSizeV2()` as the new implementation and `serializedSize()` calls one or the other based on the version. Pushing it down one level makes it easy to just delete the old function when we introduce v3. In this case the difference is not large since the logic varies only a little but it may make sense for more complex code. What do you both think?

Do you mean something like this?

```
  size_t serializedSize(IndexedVersion Version) const {
    switch (Version) {
    case Version0:
    case Version1:
      return serializedSizeV0();  // Implemented elsewhere.
    case Version2:
      return serializedSizeV2();  // Implemented elsewhere.
    }
    llvm_unreachable("unsupported MemProf version");
  }
```

If so, I like the style.  As you point out, I like the ability to easily delete an old version.  If there is a lot of code duplication between two versions, we can always factor that out to a helper function, so the code duplication shouldn't really be a big concern.

kazutakahirata commented:
I've split `serialize` and its friends into separate functions, one for each version.  Please take a look.  Thanks!

kazutakahirata committed:
Move version-specific serializers to .cpp files.

kazutakahirata commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -370,52 +396,80 @@ struct IndexedMemProfRecord {
   // mapped to functions using this GUID.
   static GlobalValue::GUID getGUID(const StringRef FunctionName);
+
+private:
```
Done.

kazutakahirata commented:
I've moved version-specific implementation to `MemProf.cpp`.  Please take a look.  Thanks!

kazutakahirata committed:
Include changes to RecordLookupTrait and RecordWriterTrait.

kazutakahirata commented:
I've included changes to `RecordLookupTrait` and `RecordWriterTrait` in the latest iteration.  Please take a look.  Thanks!

kazutakahirata committed:
Fix a typo.

teresajohnson commented on llvm/include/llvm/ProfileData/MemProf.h:
```diff
@@ -561,19 +488,21 @@ class RecordLookupTrait {
 
 // Trait for writing IndexedMemProfRecord data to the on-disk hash table.
-class RecordWriterTrait {
+template <IndexedVersion Version> class RecordWriterTrait {
```
Why pass in the version to RecordLookupTrait but template RecordWriterTrait? If it is possible to make them consistent that would be great but I might be missing the reasoning.

kazutakahirata replied:
That is because `EmitKeyDataLength` is a static function, so I wasn't able to make `Version` a member variable (as opposed to a template parameter).

Now, on a deeper look, I may be able to make `EmitKeyDataLength` a non-static member function.  I'll see if I can address that in a follow-up patch.

teresajohnson reviewed (approved):
lgtm but I do have one question below.
